clippy_lints/src/manual_arithmetic_check.rs

```diff
@@ -0,0 +1,193 @@
+    "manual arithmethic check"
+}
+
+declare_lint_pass!(ManualArithmeticCheck => [MANUAL_ARITHMETIC_CHECK]);
+
+fn is_zero(expr: &Expr<'_>) -> bool {
+    if let ExprKind::Lit(lit) = expr.peel_blocks().kind
+        && let LitKind::Int(value, _) = lit.node
+    {
+        value.get() == 0
+    } else {
+        false
+    }
+}
+
+fn get_local_var_id(expr: &Expr<'_>) -> Option<HirId> {
+    if let ExprKind::Path(QPath::Resolved(_, path)) = expr.kind
+        && let Res::Local(hir_id) = path.res
+    {
+        Some(hir_id)
+    } else {
+        None
+    }
+}
+
+struct Var {
+    span: Span,
+    hir_id: HirId,
+}
+
+impl Var {
+    fn new(expr: &Expr<'_>) -> Option<Self> {
+        get_local_var_id(expr).map(|hir_id| Self {
+            span: expr.span,
+            hir_id,
+        })
+    }
+}
+
+fn check_subtraction(
+    cx: &LateContext<'_>,
+    condition_span: Span,
+    expr_span: Span,
+    big_var: Var,
+    little_var: Var,
+    if_block: &Expr<'_>,
+    else_block: &Expr<'_>,
+) {
+    if is_zero(if_block) {
+        if is_zero(else_block) {
+            // Well, seems weird but who knows?
+            return;
+        }
+        // If the subtraction is done in the `else` block, then we need to also revert the two
+        // variables as it means that the check was reverted too.
+        check_subtraction(cx, condition_span, expr_span, little_var, big_var, else_block, if_block);
+        return;
+    }
+    let if_block = if_block.peel_blocks();
+    if is_zero(else_block)
+        && let ExprKind::Binary(op, left, right) = if_block.kind
+        && let BinOpKind::Sub = op.node
+    {
+        if Some(big_var.hir_id) == get_local_var_id(left) && Some(little_var.hir_id) == get_local_var_id(right) {
+            // This part of the condition is voluntarily split from the one before to ensure that
+            // if `snippet_opt` fails, it won't try the next conditions.
+            if let Some(big_var_snippet) = snippet_opt(cx, big_var.span)
+                && let Some(little_var_snippet) = snippet_opt(cx, little_var.span)
+            {
+                span_lint_and_sugg(
+                    cx,
+                    MANUAL_ARITHMETIC_CHECK,
+                    expr_span,
+                    "manual arithmetic check found",
+                    "replace it with",
+                    format!("{big_var_snippet}.saturating_sub({little_var_snippet})"),
+                    Applicability::MachineApplicable,
+                );
+            }
+        } else if Some(little_var.hir_id) == get_local_var_id(left)
+            && Some(big_var.hir_id) == get_local_var_id(right)
+            && let Some(big_var_snippet) = snippet_opt(cx, big_var.span)
+            && let Some(little_var_snippet) = snippet_opt(cx, little_var.span)
+        {
+            span_lint_and_then(
+                cx,
+                MANUAL_ARITHMETIC_CHECK,
+                condition_span,
+                "reverted arithmetic check before subtraction",
```

Comment:
```suggestion
                "inverted arithmetic check before subtraction",
```
sounds more correct?

Reply:
Indeed!